src/ConnectionStyle.cpp

```diff
@@ -1,233 +1,243 @@
+  return DefaultStyle;
 }
 
 
-void
+ConnectionStyle
 ConnectionStyle::
-loadJsonText(QString jsonText)
+fromJson(QString const& jsonText)
 {
-  loadJsonFromByteArray(jsonText.toUtf8());
+  auto style = defaultStyle();
+
+  style.loadJson(StyleImport::readJsonText(jsonText));
+
+  return style;
 }
 
 
-void
+// static
+QColor
 ConnectionStyle::
-loadJsonFromByteArray(QByteArray const &byteArray)
+computeNormalColor(QString const& typeId)
 {
-  QJsonDocument json(QJsonDocument::fromJson(byteArray));
+  std::size_t hash = qHash(typeId);
 
-  QJsonObject topLevelObject = json.object();
+  std::size_t const hue_range = 0xFF;
 
-  QJsonValueRef nodeStyleValues = topLevelObject["ConnectionStyle"];
+  qsrand(hash);
+  std::size_t hue = qrand() % hue_range;
 
-  QJsonObject obj = nodeStyleValues.toObject();
+  std::size_t sat = 120 + hash % 129;
 
-  CONNECTION_STYLE_READ_COLOR(obj, ConstructionColor);
-  CONNECTION_STYLE_READ_COLOR(obj, NormalColor);
-  CONNECTION_STYLE_READ_COLOR(obj, SelectedColor);
-  CONNECTION_STYLE_READ_COLOR(obj, SelectedHaloColor);
-  CONNECTION_STYLE_READ_COLOR(obj, HoveredColor);
+  return QColor::fromHsl(hue,
+                         sat,
+                         160);
+}
 
-  CONNECTION_STYLE_READ_FLOAT(obj, LineWidth);
-  CONNECTION_STYLE_READ_FLOAT(obj, ConstructionLineWidth);
-  CONNECTION_STYLE_READ_FLOAT(obj, PointDiameter);
 
-  CONNECTION_STYLE_READ_BOOL(obj, UseDataDefinedColors);
-}
+ConnectionStyle::
+ConnectionStyle()
+= default;
 
 
 QColor
 ConnectionStyle::
 constructionColor() const
 {
-  return ConstructionColor;
+  return _constructionColor;
 }
 
 
 QColor
 ConnectionStyle::
 normalColor() const
 {
-  return NormalColor;
-}
-
-
-QColor
-ConnectionStyle::
-normalColor(QString typeId) const
-{
-  std::size_t hash = qHash(typeId);
-
-  std::size_t const hue_range = 0xFF;
-
-  qsrand(hash);
-  std::size_t hue = qrand() % hue_range;
-
-  std::size_t sat = 120 + hash % 129;
-
-  return QColor::fromHsl(hue,
-                         sat,
-                         160);
+  return _normalColor;
 }
 
 
 QColor
 ConnectionStyle::
 selectedColor() const
 {
-  return SelectedColor;
+  return _selectedColor;
 }
 
 
 QColor
 ConnectionStyle::
 selectedHaloColor() const
 {
-  return SelectedHaloColor;
+  return _selectedHaloColor;
 }
 
 
 QColor
 ConnectionStyle::
 hoveredColor() const
 {
-  return HoveredColor;
+  return _hoveredColor;
 }
 
 
 float
 ConnectionStyle::
 lineWidth() const
 {
-  return LineWidth;
+  return _lineWidth;
 }
 
 
 float
 ConnectionStyle::
 constructionLineWidth() const
 {
-  return ConstructionLineWidth;
+  return _constructionLineWidth;
 }
 
 
 float
 ConnectionStyle::
 pointDiameter() const
 {
-  return PointDiameter;
+  return _pointDiameter;
 }
 
 
 bool
 ConnectionStyle::
 useDataDefinedColors() const
 {
-  return UseDataDefinedColors;
+  return _useDataDefinedColors;
+}
+
+
+void
+ConnectionStyle::
+setConstructionColor(QColor color)
+{
+  _constructionColor = std::move(color);
```

Comment:
It is copied either way

Reply:
Ah, I couldn't tell from the [documentation](http://doc.qt.io/qt-5/qcolor.html#QColor-11).

TBH, I tend to use `std::move` on anything that isn't a primitive (`int`, `long`, etc). I forget whether it has a benefit for a lot of types, so just using `std::move` means I don't have to go look up the documentation.

I'll remove these `std::move`s

Reply:
Sorry, I am not against this moves, it was just a note. No worries.

Reply:
Sounds good